### ISSUE TITLE
Nice to have: customizable bar colors

### ISSUE THREAD

**Issue by hilaryca** [Microsoft/powerbi-visuals-sankey#59]
We'd like to color code our groups so that each one is identified by the same color across all reports, making it easy for them to spot their figures in charts.

Can the Sankey allow the colors of the bars to be set too?  Ideally users would have an option to make the bars & ribbon colors match, but default would be black ribbons.

**Comment by paulbwagner** [comment]
As a generalization, it would be nice to use a table column to indicate the node color, as well as a table column as the link color. Other sankey diagrams such as those in D3 or Python libraries have these sorts of properties driven by the data and not by fixed format settings, which gives the user maximum flexibility.

**Comment by lrlucchini** [comment]
Yes please. This has me going through the code here trying to see if there's an easy way to replicate the process to color the links, but doing that for the nodes instead. In the configuration/capabilities JSON file it already specifies a "Node" section in the format pane, identically to the "Links" section which allows you to change the color of each link. I think the original intent was to allow us to change the nodes, but somewhere in the process that got removed/broken. I am trying to create my own custom visual to fix for this by copying their code, but I am not sure what I'm doing at this point.

**Comment by jeremyfirth** [comment]
> Yes please. This has me going through the code here trying to see if there's an easy way to replicate the process to color the links, but doing that for the nodes instead. In the configuration/capabilities JSON file it already specifies a "Node" section in the format pane, identically to the "Links" section which allows you to change the color of each link. I think the original intent was to allow us to change the nodes, but somewhere in the process that got removed/broken. I am trying to create my own custom visual to fix for this by copying their code, but I am not sure what I'm doing at this point.

I'm in the same boat. I have a little coding background, and I was relieved to find the source code for this visualization on github. Have you made any progress with being able to color-code the nodes manually? I'm looking for that function specifically.

**Comment by helen508light** [comment]
Thank you for your feedback.
We'll consider your suggestions in the next versions of the visual.

@jeremyfirth
There are no implemented features you had mentioned here for now.  
Please note that source code for the visual is public, feel free to fork and modify it for your needs.

**Comment by harry843** [comment]
I agree with the comments above. I would find it hugely beneficial for the colouring of node bars to be a feature included in the next release. I am also new to coding and unsure how to go about doing this manually using the source code on github.

**Comment by mneira10** [comment]
+1

**Comment by ncardona10** [comment]
It would be incredibly usefull !!

**Comment by joespiff** [comment]
I've made some changes to the code in my fork of this repo in an effort to enable this feature.  It is still experimental, but I'd love to hear feedback from anyone who tries it.  If I can get some validation that it is solid, I may make a pull request to get it incorporated into the official repo.

**Comment by zBritva** [comment]
@joespiff It's good that community wants to help to improve the visual. Thanks!

But the problem is:

Power BI binds properties to data (data rows). In this case it's color.
For 1 row you can bind one instance of color. (for one object)
For 5 row you can bind 5 instances of color. (for one object)
e.t.c

**Details:**
We have data:

| TARGET | SOURCE | VALUE |
|-|-|-
| A| B| 1

There is one row. And we can bind only one color per row.
It means if you have only one object, you can store only one color (for source node or for target node).

There we have a bug:
![sankey_nodes_color](https://user-images.githubusercontent.com/10897951/90829199-fe7cdb00-e347-11ea-8f35-a88a3a6b1f0b.gif)

It's limitation of [categorical data view mapping](https://docs.microsoft.com/en-us/power-bi/developer/visuals/dataview-mappings#categorical-data-mapping). But it requires rework code of sankey to switch from categorical to matrix data view mapping

Probably it can be resolved by using [matrix data view mapping](https://docs.microsoft.com/en-us/power-bi/developer/visuals/dataview-mappings#matrix-data-mapping), but I am not sure.

Another tricky options is: use two different objects: one for source nodes color another for target nodes color. 
Disadvantage is, you can assign different colors in source and target for the same node. It can confuse the user.

I hope I explained the limitations. Ask if you have questions.

**Comment by Mike-Honey** [comment]
> I've made some changes to the code in my fork of this repo in an effort to enable this feature. It is still experimental, but I'd love to hear feedback from anyone who tries it. If I can get some validation that it is solid, I may make a pull request to get it incorporated into the official repo.

@joespiff - I'm keen to try your fork, but I cant find it?

**Comment by Mike-Honey** [comment]
Please implement this by adding input columns for the source and destination node colors, and for the link color.  That's 3 input columns in all.

Trying to achieve consistent colors with this visual is currently a nightmare - it resets node colors seemingly at random.
